src/Compilers/CSharp/Portable/Symbols/MemberSignatureComparer.cs

```diff
@@ -239,7 +239,7 @@ internal class MemberSignatureComparer : IEqualityComparer<Symbol>
         public static readonly MemberSignatureComparer LambdaReturnInferenceCacheComparer = new MemberSignatureComparer(
             considerName: false,                // valid invoke is always called "Invoke"
             considerExplicitlyImplementedInterfaces: false,
-            considerReturnType: false,          // do not care
+            considerReturnType: true,          // to differentiate Task types
```

Comment:
In C#6 when you infer the return type of a lambda `async (x) => x+1` when matched against a delegate type e.g. `Func<int,Task<T>>`, it takes into account the parameter types of the delegate (in this case `x:int`), and infers the type of the body of the lambda (in this case `int`), and the presence of the async modifier means the inferred return type is `Task<int>`. It just sticks `Task` in there blindly for any async lambda.

The compiler caches the result of this computation. So, if we try to infer the return type of the lambda given any delegate type, so long as the new delegate type has the same parameter types as the old one, then C#6 can simply look it up in the cache. The cache was keyed solely off the parameter types.


In C#7, if you inferred the return type of `async (x) => x+1` in the context of a target delegate `Func<int,Task<T>>` then it would infer the return type `Task<int>`. But if you inferred its return type in the context of a target delegate `Func<int,ValueTask<T>>` then it would infer the return type `ValueTask<int>`.

Thus, the new computation depends on the delegate's return type as well as its parameter types.

Therefore the cache must key off return type as well as parameter types.